tiledb/sm/query/readers/reader_base.h

```diff
@@ -206,7 +207,7 @@ class ReaderBase : public StrategyBase {
    * Boolean, per fragment, to specify that we need to load timestamps for
    * deletes. This matches the fragments in 'fragment_metadata_'
    */
-  std::vector<bool> timestamps_needed_for_deletes_;
+  std::vector<bool> timestamps_needed_for_deletes_and_updates_;
```

Comment:
We have an implicit synchronization requirement between three different vectors here, which complicates the code in various places, and is a potential source of bugs.

One possibility to improve this would be a small class to encapsulate all three concepts, and only store `std::vector<MaybeTimestampedQC>`. Something like this:

```
class MaybeTimestampedQC {
  QueryCondition QC;
  std::optional<tiledb_timestamp_t> timestamp;
  bool timestamp_needed;

  const QueryCondition& qc();
  const QueryCondition& timestamped_qc(); // generates the timestamped version on-demand
  ...
};
```